extras/bmp280/bmp280.c

```diff
@@ -46,116 +46,155 @@
 #define BMP280_REG_CONFIG      0xF5 /* bits: 7-5 t_sb; 4-2 filter; 0 spi3w_en */
 #define BMP280_REG_CTRL        0xF4 /* bits: 7-5 osrs_t; 4-2 osrs_p; 1-0 mode */
 #define BMP280_REG_STATUS      0xF3 /* bits: 3 measuring; 0 im_update */
+#define BMP280_REG_CTRL_HUM    0xF2 /* bits: 2-0 osrs_h; */
 #define BMP280_REG_RESET       0xE0
 #define BMP280_REG_ID          0xD0
 #define BMP280_REG_CALIB       0x88
+#define BMP280_REG_HUM_CALIB   0x88
 
 
-#define BMP280_CHIP_ID         0x58 /* BMP280 has chip-id 0x58 */
 #define BMP280_RESET_VALUE     0xB6
 
-typedef struct __attribute__((packed)) {
-    uint16_t dig_T1;
-    int16_t  dig_T2;
-    int16_t  dig_T3;
-    uint16_t dig_P1;
-    int16_t  dig_P2;
-    int16_t  dig_P3;
-    int16_t  dig_P4;
-    int16_t  dig_P5;
-    int16_t  dig_P6;
-    int16_t  dig_P7;
-    int16_t  dig_P8;
-    int16_t  dig_P9;
-} BMP280_Calib;
-
-static BMP280_Calib calib_data;
 
 void bmp280_init_default_params(bmp280_params_t *params)
 {
     params->mode = BMP280_MODE_NORMAL;
     params->filter = BMP280_FILTER_OFF;
     params->oversampling = BMP280_STANDARD;
+    params->oversampling_humidity = BMP280_STANDARD;
     params->standby = BMP280_STANDBY_250;
 }
 
-static uint8_t read_register8(uint8_t addr)
+static bool read_register8(uint8_t i2c_addr, uint8_t addr, uint8_t *value)
 {
-    uint8_t r = 0;
-    if (!i2c_slave_read(BMP280_ADDRESS, addr, &r, 1)) {
-        r = 0;
-    }
-    return r;
+    return i2c_slave_read(i2c_addr, addr, value, 1);
 }
 
-/**
- * Even though value is signed the actual value is always positive.
- * So, no need to take care of sign bit.
- */
-static bool read_register24(uint8_t addr, int32_t *value)
+static bool read_register16(uint8_t i2c_addr, uint8_t addr, uint16_t *value)
 {
-    uint8_t d[] = {0, 0, 0};
-    if (i2c_slave_read(BMP280_ADDRESS, addr, d, sizeof(d))) {
-        *value = d[0];
-        *value <<= 8;
-        *value |= d[1];
-        *value <<= 4;
-        *value |= d[2]>>4;
+    uint8_t d[] = {0, 0};
+    if (i2c_slave_read(i2c_addr, addr, d, sizeof(d))) {
+        *value = d[0] | (d[1] << 8);
         return true;
     }
     return false;
 }
 
-static bool check_chip_id()
+static bool read_calibration_data(bmp280_calib_t *calib_data)
 {
-    return (read_register8(BMP280_REG_ID)==BMP280_CHIP_ID);
+    uint8_t i2c_addr = calib_data->i2c_addr;
+
+    if (read_register16(i2c_addr, 0x88, &calib_data->dig_T1) &&
```

Comment:
I would still prefer to read calibration data as a one block of data into "packed" struct. At least pressure calibration data as those registers goes one after another. Unfortunately it can't be done for humidity calibration data.